File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} list1
# @param {ListNode} list2
# @return {ListNode}

def merge_two_lists(list1, list2)
    dummy = ListNode.new
    tail = dummy 
    
    while list1 and list2 
        if list1.val < list2.val
            tail.next = list1
            list1 = list1.next
        else 
            tail.next = list2
            list2 = list2.next
        end 
         tail = tail.next 
    end 
   
    if list1 
        tail.next = list1
    elsif list2
        tail.next = list2
    end
    dummy.next
    
    
end

File: 21-merge-two-sorted-lists/21-merge-two-sorted-lists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function(list1, list2) {
    let dummyHead = new ListNode(null)
  let tail = dummyHead
  let curr1 = list1
  let curr2 = list2
  
  while (curr1 !== null && curr2 !== null){
    if(curr1.val < curr2.val){
      tail.next = curr1
      curr1 = curr1.next 
    } else{
      tail.next = curr2
      curr2 = curr2.next
    }
    tail = tail.next
  }
  
  if(curr1 !== null) tail.next = curr1
  if(curr2 !== null) tail.next = curr2
  
  
  return dummyHead.next
    
};